drivers: ieee802154: cc13xx/cc26xx_subg radio configuration

fgrandel opened the pull request:
This change set cleans up the CC13xx/CC26xx Sub-GHz driver and introduces some fixes and preparatory changes for multi-PHY support (multiple channel pages). The driver still has some issues wrt PHY standard compliance which have been documented (see the corresponding commit comments). These cannot be removed w/o breaking backwards compatibility and will therefore need to be maintained as a "proprietary" PHY for a while.

UPDATE: **This change set also works around a hardware bug that wouldn't allow any link between CC13/26xxP (e.g. CC1352P1 LaunchPad) and CC13/26xxR (e.g. LPSTK-CC1352R SensorTag) devices at all, see "Known Issues" in [TI's SDK release notes](https://software-dl.ti.com/simplelink/esd/simplelink_cc13xx_cc26xx_sdk/6.41.00.17/exports/release_notes_simplelink_cc13xx_cc26xx_sdk_6_41_00_17.html#known-issues).** (Both bug and solution were confirmed in hardware.)

This change is part of a series of changes required to enable TSCH on Zephyr, see #50336. The relevance of supporting channel page 9 as prepared in this PR is that it provides multi-channel support in the 863 MHz band which is required for channel hopping in TSCH.

All changes have been extensively regression tested against a combination of two CC1352P1 LaunchPad dev boards (+ some additional test coverage on CC1352R). A full set of measurement results can be seen in https://github.com/zephyrproject-rtos/zephyr/pull/58439#issuecomment-1585746656 below.

Commit:
drivers: ieee802154: cc13/26xx_subg: clean up constants

The CC13xx/CC26xx Sub-GHz driver header file defined several constants
that were not used in the driver.

Other constants could be replaced with generic constants which were
introduced in the prior commit.

This change removes and/or replaces redundant definitions.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: remove redundant configuration

The low-level configuration of the chip's radio commands was mostly
redundant. This change removes redundant configuration code.

This is also relevant as a preparation to supporting further frequency
bands and operating modes on the same SUN FSK channel page with
similar but slightly different settings (center frequencies, channel
spacing, modulation index as defined in the standard). The SUN FSK
standard defines plenty of such variations with different physical
characteristics and trade-offs. Such variations are highly relevant in
industrial applications which will be targeted by TSCH. Using the
correct settings is required for additional features (e.g. frequency
hopping) and interoperability.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: fix invalid KConfig reference

The driver contained references to KConfig variables w/o the required
CONFIG_ prefix. This change introduces the missing prefixes.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: readability improvements

This change introduces standard variable names used elsewhere in the
stack for improved naming consistency and readability.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: fix/document non-standard CSMA/CA

The CC13xx/CC26xx Sub-GHz driver announces a hardware CSMA/CA capability
which it provides only partially. This change documents the gap.

The change also fixes two related issues with the current CCA
implementation:

- The given default ED threshold was above the allowed threshold
  defined in the specification.

- The CCA timeout was not calculated according to the requirements
  defined in the standard.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: remove dead code

The channel-to-frequency conversion had unreachable code which is
removed in this change.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: inline documentation

This change introduces inline documentation with references to the
current version of the IEEE 802.15.4 standard.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

cfriedt commented:
This looks great - thanks @fgrandel !

@vaishnavachath - will you verify some of the radio settings? Particularly in `drivers: ieee802154: cc13/26xx_subg: remove redundant configuration`? 

Overall, very nice improvements.

I have 2 requests:
1. Can you run `samples/net/sockets/echo_server` in UDP mode for e.g. a 24 hour period and report packet loss? Before and after your change would be ideal, but I know that is a lot to ask
2. I don't believe there is an overlay file, but it would be handy to run `samples/net/zperf/` for some amount of time and report the output?

fgrandel commented:
Thanks @cfriedt for your review, encouragement and helpful feedback! More is to follow in future PRs soon as indicated by the TODO tags in the PR - especially improvements in standard conformance (interoperability and CSMA/CA timing) and multi-channel SubG PHYs from the SUN range.

> * Can you run `samples/net/sockets/echo_server` in UDP mode for e.g. a 24 hour period and report packet loss? Before and after your change would be ideal, but I know that is a lot to ask

Very good idea, will do. As long as this was not successfully tested on real hardware, I'll leave this PR in the draft state.

> * I don't believe there is an overlay file, but it would be handy to run `samples/net/zperf/` for some amount of time and report the output?

Again - very good idea, will do. I'll report results back in this PR.

The biggest stumbling block wrt testing with our dev kits (CC1352P1 Launch Pad and CC1352R SensorKit) right now is debugger support - both XDS and JLink seem to be at odds with the software interrupts caused by the radio stack - at least when run through Zephyr's debugging infrastructure (west debug). Whenever an interrupt occurs (which means almost on each step) code execution is halted somewhere in the TI IRQ handling stack and the next line of code will never be reached.

Maybe @vaishnavachath can help me with that or someone else from TI who knows the proprietary ROM radio stack? Problem is that similar topics reported on TI E2E were not (yet) answered plus TI engineers on that forum seem to be a little reluctant to reproduce problems on the Zephyr stack. ;-)

It would also be very helpful to get some insight from TI how the proprietary radio's Carrier Sense command parameters are to be interpreted. The documentation, existing stacks and samples in the SDK are a bit contradicting in that respect.

cfriedt commented:
@fgrandel - if you are open to contract work, some of the things you've mentioned would likely be of some interest to TI. At one point, I was trying to get BLE working in Zephyr (again with the ROM drivers), but TI was not open to a contract at that time. Perhaps they've chang d their minds at this point. Of course, every company has policy, so it's hard to say if it's feasible, but I mostly do like TI's product, and it would be amazing to have better radio integration in Zephyr.

fgrandel commented:
> If you are open to contract work

Thanks for that kind offer! I'm currently under contract but maybe there's some room for co-operation. I'll send you a PM on discord to explain my situation.

> I mostly do like TI's product though, and it would be amazing to have better radio integration in Zephyr.

Totally agreed. I'm willing to invest quite a bit as we'd love to base a product on top of these SoCs.

> On a side note, I have also removed the jumper connecting the XDS to the cc and have successfully used OpenOCD in the past directly. It's been a while though.

Yep - we tried all options, including that one. We spent the day looking into the debugger problem, both via OpenOCD and JLink (Segger Ozone and M-Cortex extension in VSCode). The problem could be narrowed down to radio operations triggered by software interrupts running in the background. These software interrupts seem to trigger while stepping although both OpenOCD and JLink integrations claim to mask interrupts when stepping. If we set PRIMASK=1 manually the interrupts really do not trigger. Maybe the debuggers no longer use PRIMASK nowadays but some other less intrusive masking strategy and the radio MCU (which the debugger does know nothing about) somehow re-activates interrupts by the same mechanism while we're stepping? No idea...

Our workaround now is to ensure that we only set breakpoints in places where we are sure that no background operation is active. Not ideal but ok until we find a better solution.

Commit:
drivers: ieee802154: cc13/26xx_subg: fix CCA method

The driver's CCA method had various issues and would always return an
error code. This is fixed in this change.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: improve CSMA/CA compliance

Switch the driver to the soft CSMA/CA algorithm as an intermediate
compromise for improved standard compliance (namely expontential
backoff) until true hardware support can be implemented by chaining
radio commands.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
@cfriedt The CSMA/CA mechanism has now been converted to soft CSMA/CA for improved standard conformance like we discussed yesterday. I tested this approach on real hardware by interferring with a continuous carrier signal from a second device which worked as expected. This also preliminarily confirms the CSMA/CA fixes underneath this change set. The other hardware tests you proposed are outstanding and will be executed asap.

fgrandel commented:
@niflostancu Tagging you because you offered in #46050 to help with hardware exposure of changes re CC13/26x2 platforms. If you like you can have a go with this branch. Would be very appreciated. You could do testing with your own set-up or help with the test scenarios that @cfriedt proposed (see his comment above for more detailed requirements):
1. Checking PER for 24 hours before/after this change.
2. Create support for zperf and execute it against these changes before/after.

We'll do the same but redundancy is a good thing when it comes to testing. :-)

fgrandel commented:
@niflostancu Tagging you because you offered in #46050 to help with hardware exposure of changes re CC13/26x2 platforms. If you like you can have a go with this branch. Would be very appreciated. You could do testing with your own set-up or help with the test scenarios that @cfriedt proposed (see his comment above for more detailed requirements):
1. Checking PER for 24 hours before/after this change.
2. Create support for zperf and execute it against these changes before/after.

We'll do the same but redundancy is a good thing when it comes to testing. :-)

vaishnavachath commented:
@fgrandel Thank you for your contributions, overall the PR looks great. 
@cfriedt @fgrandel , Sure I will verify the changes on my setup and am reviewing the changes now, have been out of office last two weeks due to an injury.

@fgrandel On the ROM radio stack, I have requested someone from the Connectivity team to help here, since I am also from a different product line I also don't have access to the necessary information you need.

fgrandel commented:
> have been out of office last two weeks due to an injury.

@vaishnavachath I'm really sorry to hear that. :-( Thanks for getting someone from the connectivity team involved.

niflostancu commented:
> @niflostancu Tagging you because you offered in https://github.com/zephyrproject-rtos/zephyr/pull/46050 to help with hardware exposure of changes re CC13/26x2 platforms. If you like you can have a go with this branch. Would be very appreciated. You could do testing with your own set-up or help with the test scenarios that @cfriedt proposed (see his comment above for more detailed requirements):

Hello.
Of course, I will get on it later this week.

Commit:
drivers: ieee802154: cc13/26xx_subg: remove unused radio commands

The CMD_CLEAR_RX and CMD_SET_TX_POWER commands are declared and
initialized but not used anywhere. They are therefore removed to reduce
RAM/flash footprint.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
boards: cc1352p1_launchxl: fix antenna mux config

The antenna MUX configuration should explicitly define PULL states as
giving no value will leave the GPIO register in an unspecified state.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: conditional RX after set_channel

The call to radio_api->set_channel() must not switch on RX if it was
previously off.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: fix header len const

The length field in the header refers to the size of the MAC so it
shouldn't rely on constants describing PHY header length. While
currently both constants have the same value this will no longer be true
for enhanced PHYs and/or MAC frames as the number of FCS bytes may then
be four.

Also introduces an assertion that ensures that the given package buffer
does not exceed the TX buffer's length. An assertion is enough as the
package buffer is allocated at compile time.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: fix subg_start() return value

The radio API should indicate errors when the interface is started.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13/26xx_subg: fix PHY overrides

PHY overrides have been checked against the latest version of TI's
SmartRF(TM) Studio. The result was regression tested (PER/performance)
against LAUNCHXL-CC1352P1 boards.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
@niflostancu @vaishnavachath @cfriedt I did a first regression test with the echo client/server. Further regression tests (including baseline tests with the current mainline) follow and will be added in this same comment so that you don't get spammed all the time when a new result arrives.

The result table calculates PER based on IPv6 packages as `net stats` does not report IEEE 802.15.4 frames/fragments. So the actual L1 PER rate should be considerably less as the echo client/server causes a lots of fragmentation due to huge payloads. I worked with very low TX Power as such tests are otherwise problematic from a duty cycle perspective under German Federal Law. That's also the reason why I used a patched version of the echo client that inserts pauses between packets on the TX side.

| Test conditions | Protocol | Sent Pkts | Received Pkts | Dropped Pkts | L3 PER |
|---------------------|-------------|---------------|--------------------|--------------------|------------|
| -20 dBm TX Pwr, 1m distance, no ACK | UDP | 137399 | 136542 | 0 |  6.24‰ |

fgrandel commented:
Driver maintainers, @cfriedt @vaishnavachath @niflostancu A full set of regression tests (echo client/server + zperf) is now available. Can you confirm these typical results on your hardware? It would be especially nice to see other hardware altogether being tested by someone.

cfriedt commented:
@jadonk - check out these numbers - really fantastic work here on SubG (may need to expand some older comments).

Commit:
drivers: ieee802154: cc13/26xx_subg: R-to-P link workaround

A known issue exists that does not allow a CC13/26x2R device to
establish a link to a CC13/26x2P device unless the capacitor array is
tuned to a non-default value in the P device.

See SimpleLink(TM) cc13xx_cc26xx SDK 6.20+ Release Notes, Known Issues.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

niflostancu commented:
Okay, so I finally tested the PR using two LaunchXL 1352Ps, echo client/server with the client set at 20dBm (max :D) TX power & 100ms interval between the UDP packets for ~20h (~10m in separate rooms with solid walls, ACK disabled):
```
client:~$ net stats
Interface 0x200013a8 (IEEE 802.15.4) [1]
========================================
IPv6 recv      319009   sent    320295  drop    1271    forwarded       0
IPv6 ND recv   4698     sent    4937    drop    0
IPv6 MLD recv  0        sent    3       drop    0
IP vhlerr      0        hblener 0       lblener 0
IP fragerr     0        chkerr  0       protoer 0
UDP recv       313040   sent    315347  drop    1271
UDP chkerr     0
Bytes received 45265261
Bytes sent     44543737
Processing err 10719
```

This gives a PER of 0.73%, which is pretty nice.
I also tested it on my company's custom hardware (battery based sensor nodes + gateway) and no problems were found.
Note: I only tested this PR, not the mainline, but the results should be valid.

I plan next week to put dozen of battery-powered wireless nodes to periodically (~10s) transmit their data over a noisy environment (I can generate air traffic using my LaunchXL) and record their packets sent/received stats, but this requires some extra coding so it might take a while to do.

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:
```diff
@@ -154,33 +137,55 @@ static uint32_t rf_prop_overrides_tx_20[] = {
+	.txPower = 0x013f, /* 14 dBm, see TRM 25.3.3.2.16 */
+	.centerFreq = 906, /* Set channel page zero, channel 1 by default, see IEEE 802.15.4,
+			    * section 10.1.3.3.
+			    * TODO: Use compliant SUN PHY frequencies from channel page 9.
```
Just for clarification, is there already an issue (or perhaps draft PR) in place to cover this?

fgrandel replied:
No, not yet, just the generic roadmap in #50336. This is a mandatory pre-condition for the "CH" (aka channel hopping) part in the TSCH acronym.

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:
```diff
@@ -258,26 +263,44 @@ static inline int ieee802154_cc13xx_cc26xx_subg_channel_to_frequency(
 		return -EINVAL;
 	}
 
+	/* TODO: This incorrectly mixes up legacy O-QPSK SubGHz PHY channel page zero
```
IIRC, this was the only way to make it work with wpanusb.

fgrandel replied:
The wpanusb code seems to be limited to channel page zero but then uses the channel number abstraction so it should be independent of low-level PHY settings (including but not limited to the modulation scheme). The original code by 
[finikorg](https://github.com/finikorg/wpanusb) has a hard coded channel list, [@jadonk's fork](https://github.com/jadonk/wpanusb) fixes that for the supported channel list but still seems to have channel page zero hard coded.

wpanusb allows for a channel page being set in set_channel() but it doesn't look as if this also updated the supported channels, so probably setting a non-zero channel page will break the supported channels which again will let the kernel driver bail out if trying to set a channel above 10 (which is totally valid on channel page 9/10 for example). This means that probably wpanusb will not work with channel pages 9 or 10 (SUN FSK) but continues to work with channel page zero (which must not be removed anyway).

The problem is that CC13/26xx doesn't support channel page zero in the subG area at all as it doesn't support O-QPSK subG modulation. Therefore there's no way to fix the PHY for this driver for channel page zero. This means that "fixing" in this case only means additionally supporting channel page 9 but leaving channel page zero exactly as it is (including its non-standard PHY).

Hope that explains a bit what that change is going to be about?

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:
```diff
@@ -339,30 +362,41 @@ static enum ieee802154_hw_caps
+		return -EIO;
+	}
+
+	/* TODO: Should we re-enable RX? Would not normally
```
Proabably would be safe to do. I can't remember the exact details. @vaishnavachath ?

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:
```diff
@@ -419,20 +453,17 @@ static int ieee802154_cc13xx_cc26xx_subg_set_channel(
 		goto out;
 	}
 
-	/* Run BG receive process on requested channel */
```
I'm not 100% sure it's safe to remove this

fgrandel replied:
In any case leaving RX on after setting the channel would be a bug as most of the time the radio is off when you set the channel and then leaving RX on would drain the battery. That's why I removed it.

Setting the channel must be done while not associated and scanning deals with switching the device on (aka going into RX mode) - so in practice it shouldn't be a problem.

But I agree that it would be safer to check whether RX is on at the beginning of the CCA routine (looking at the cmd's status == 2) and switch it back on if it was on before and leave it off otherwise. I'll fix that.

fgrandel replied:
@cfriedt Can you have a look again? I think I fixed this now.

vaishnavachath commented:
@fgrandel @cfriedt I did overnight Echo Client/Server testing on 2 BeagleConnect Freedom, 0 dBm TX Pwr, 10m distance and saw a PER of 0.21% ACK disabled, did not test mainline separately.
Also noting that will need to fix the `if defined(CONFIG_SOC_CC1352R)` and `if defined(CONFIG_SOC_CC1352P)` to include the corresponding R7/P7 SoC variants as well, I fixed it up locally for testing BCF, will submit a separate PR for that.

I will test wpanusb/bcfserial functionality as well today and confirm the results.

fgrandel commented:
> PER of 0.21% ACK disabled

Nice, thank you for testing that! :-)

> Also noting that will need to fix the `if defined(CONFIG_SOC_CC1352R)` [...] will submit a separate PR for that.

Ok, I'll do nothing about it then.

fgrandel commented:
@vaishnavachath Could you reset to the latest version of this branch before continuing to test? @cfriedt had some good remarks which I turned into code so it would be good to test the actual thing. It would also be good to have results with ACK + CSMA/CA switched on as these are the areas that should show most improvement in PER as compared to mainline.

fgrandel commented:
@cfriedt @vaishnavachath @niflostancu Thanks for your support. This was a really nice community team work experience. :-)
